src/modules/linux-micro/kmod/kmod.c

```diff
@@ -0,0 +1,474 @@
+SOL_LOG_INTERNAL_DECLARE_STATIC(_log_domain, "linux-micro-kmod");
+
+#include "sol-file-reader.h"
+#include "sol-mainloop.h"
+#include "sol-platform-linux-micro.h"
+#include "sol-util.h"
+
+struct kmod_data {
+    struct kmod_ctx *kmod;
+    struct {
+        int fd;
+        struct sol_fd *watch;
+    } uevent;
+};
+
+static struct kmod_data context;
+
+static int
+kmod_apply_value(struct kmod_ctx *kmod, char *alias)
+{
+    const int probe_flags = KMOD_PROBE_APPLY_BLACKLIST;
+    struct kmod_list *itr, *modlist = NULL;
+    int r = 0;
+
+    SOL_DBG("Trying to load module for alias: %s", alias);
+
+    r = kmod_module_new_from_lookup(kmod, alias, &modlist);
+    if (r < 0) {
+        SOL_ERR("Failed to lookup alias '%s': %s", alias,
+            sol_util_strerrora(errno));
+        return r;
+    }
+
+    if (!modlist) {
+        SOL_INF("No modules found for alias: '%s'", alias);
+        return -ENOENT;
+    }
+
+    kmod_list_foreach(itr, modlist) {
+        struct kmod_module *mod;
+        int state, err;
+
+        mod = kmod_module_get_module(itr);
+        state = kmod_module_get_initstate(mod);
+
+        switch (state) {
+        case KMOD_MODULE_BUILTIN:
+            SOL_DBG("Module '%s' is builtin", kmod_module_get_name(mod));
+            break;
+
+        case KMOD_MODULE_LIVE:
+            SOL_DBG("Module '%s' is already loaded", kmod_module_get_name(mod));
+            break;
+
+        default:
+            err = kmod_module_probe_insert_module(mod, probe_flags,
+                NULL, NULL, NULL, NULL);
+
+            if (err == 0)
+                SOL_INF("Inserted module '%s'", kmod_module_get_name(mod));
+            else if (err == KMOD_PROBE_APPLY_BLACKLIST)
+                SOL_INF("Module '%s' is blacklisted", kmod_module_get_name(mod));
+            else {
+                SOL_WRN("Failed to insert '%s': %s", kmod_module_get_name(mod),
+                    sol_util_strerrora(errno));
+                r = err;
+            }
+        }
+
+        kmod_module_unref(mod);
+    }
+
+    kmod_module_unref_list(modlist);
+    return r;
+}
+
+static int
+kmod_apply_file(struct kmod_ctx *kmod, int fd)
+{
+    int err = 0, ret;
+    struct sol_file_reader *reader;
+    struct sol_str_slice content = SOL_STR_SLICE_EMPTY;
+    struct sol_str_slice *itr;
+    uint16_t idx;
+    struct sol_vector lines;
+
+    reader = sol_file_reader_fd(fd);
+    if (!reader) {
+        SOL_ERR("Could not open config file");
+        return err;
+    }
+
+    content = sol_file_reader_get_all(reader);
+    if (content.len == 0)
+        goto finish;
+
+    lines = sol_util_str_split(content, "\n", 0);
+    SOL_VECTOR_FOREACH_IDX (&lines, itr, idx) {
+        char *alias = NULL;
+
+        sol_str_slice_remove_leading_whitespace(itr);
+        if (*itr->data == '#' || *itr->data == ';' || *itr->data == '\0')
+            continue;
+
+        sol_str_slice_remove_trailing_whitespace(itr);
+        alias = strndupa(itr->data, itr->len);
```

Comment:
indeed, using `alloca()` and variations will explode stack if called inside the loop... the stack just goes back once the loop is finished :-/

I do recommend you to use `struct sol_buffer` here, initialize it before the loop and then call `sol_buffer_set_slice()`, it will reuse the allocated memory if possible, otherwise will grow as needed. Remember to call `sol_buffer_fini()` afterwards to release allocated memory.

by default, the buffer adds null-terminator, so safe to use as a string.